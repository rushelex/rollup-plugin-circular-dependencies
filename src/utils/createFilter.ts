import { isAbsolute, posix, win32 } from 'node:path';
import { cwd } from 'node:process';
import pm from 'picomatch';

// A valid filter pattern for picomatch
export type FilterPattern = ReadonlyArray<string | RegExp> | string | RegExp | null;

function ensureArray(thing: any) {
  if (Array.isArray(thing)) {
    // eslint-disable-next-line ts/no-unsafe-return
    return thing;
  }

  if (thing == null) {
    return [];
  }

  // eslint-disable-next-line ts/no-unsafe-return
  return [thing];
}

const normalizePathRegExp = new RegExp(`\\${win32.sep}`, 'g');
function normalizePath(filename: string) {
  return filename.replace(normalizePathRegExp, posix.sep);
}

function getMatcherString(id: string) {
  if (isAbsolute(id) || id.startsWith('**')) {
    return normalizePath(id);
  }
  // resolve('') is valid and will default to process.cwd()
  const basePath = normalizePath(cwd())
    // escape all possible (posix + win) path characters that might interfere with regex
    .replace(/[-^$*+?.()|[\]{}]/g, '\\$&');
  // Note that we use posix.join because:
  // 1. the basePath has been normalized to use /
  // 2. the incoming glob (id) matcher, also uses /
  // otherwise Node will force backslash (\) on windows
  return posix.join(basePath, normalizePath(id));
}

export function createFilter(include: FilterPattern, exclude: FilterPattern) {
  const getMatcher = (id: string | RegExp) => id instanceof RegExp
    ? id
    : {
        test: (what: string) => {
          // this refactor is a tad overly verbose but makes for easy debugging
          const pattern = getMatcherString(id);

          const fn = pm(pattern, { dot: true });
          return fn(what);
        },
      };

  const includeMatchers = ensureArray(include).map(getMatcher);
  const excludeMatchers = ensureArray(exclude).map(getMatcher);

  if (!includeMatchers.length && !excludeMatchers.length) {
    return (id: any) => typeof id === 'string' && !id.includes('\0');
  }

  return function result(id: any) {
    if (typeof id !== 'string') {
      return false;
    }

    if (id.includes('\0')) {
      return false;
    }

    const pathId = normalizePath(id);

    for (let i = 0; i < excludeMatchers.length; ++i) {
      const matcher = excludeMatchers[i];
      if (matcher.test(pathId)) {
        return false;
      }
    }

    for (let i = 0; i < includeMatchers.length; ++i) {
      const matcher = includeMatchers[i];
      if (matcher.test(pathId)) {
        return true;
      }
    }

    return !includeMatchers.length;
  };
}
