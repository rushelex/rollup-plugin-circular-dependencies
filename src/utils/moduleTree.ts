import type { PluginContext } from 'rollup';

import type { Context } from '../context';
import type { ModuleNode } from '../module';

export function generateModuleTree(ctx: Context, rootModuleNode: ModuleNode) {
  const createdModuleNode = new Set<string>();
  const moduleNodes = ctx.moduleNodes;

  function getModuleChildNodes(node: ModuleNode): ModuleNode[] {
    return node.importedIds.reduce<ModuleNode[]>((acc, moduleId) => {
      const moduleNode = moduleNodes.get(moduleId);

      if (moduleNode) {
        acc.push(moduleNode);
      }

      return acc;
    }, []);
  }

  function recursionBuild(node: ModuleNode) {
    // For each traversal, first put yourself into the generated set to avoid referencing your own scene.
    createdModuleNode.add(node.id);

    const childNodes = getModuleChildNodes(node);

    for (const childNode of childNodes) {
      // The current subtree is not generated, then recursion
      if (!createdModuleNode.has(childNode.id)) {
        recursionBuild(childNode);
      }

      node.children.add(childNode);
    }
  }

  recursionBuild(rootModuleNode);
}

export function generateCycleNodesMap(ctx: Context, rootModuleNode: ModuleNode, pluginContext: PluginContext) {
  // Used to store the last traversed nodes with circular dependencies.
  const cycleNodesMap = new Map<string, ModuleNode[]>();

  // Save the nodes that have been DFSed to avoid repeated traversal
  const visitedNodeIds = new Set<string>();

  // Save the unique node ids to call onDetected
  const detectedNodeIds = new Set<string>();

  function depthFirstTraversal(node: ModuleNode, visitedPaths: Set<string>) {
    // Has been visited and ends directly
    if (visitedNodeIds.has(node.id)) {
      return;
    }

    // First add yourself to the path to avoid situations where you reference yourself.
    visitedPaths.add(node.id);

    for (const childNode of node.children) {
      if (visitedPaths.has(childNode.id)) {
        if (!detectedNodeIds.has(childNode.id)) {
          ctx.options.onDetected(childNode.id, pluginContext);
        }

        detectedNodeIds.add(childNode.id);

        generateAndInsertCycleNodes(childNode, visitedPaths, cycleNodesMap);
      }
      else {
        depthFirstTraversal(childNode, visitedPaths);
      }
    }

    visitedNodeIds.add(node.id);

    // Pop the current node out of the path
    visitedPaths.delete(node.id);
  }

  depthFirstTraversal(rootModuleNode, new Set());

  return cycleNodesMap;
}

function generateAndInsertCycleNodes(
  node: ModuleNode,
  visitedPaths: Set<string>,
  cycleNodesMap: Map<string, ModuleNode[]>,
) {
  const cycleNodes: ModuleNode[] = [];

  let currentNode: ModuleNode | undefined = node;

  do {
    cycleNodes.push(currentNode);

    if (currentNode.children.size === 0) {
      break;
    }

    currentNode = Array.from(currentNode.children).find((item) => visitedPaths.has(item.id));
  } while (currentNode && currentNode !== node);

  const sortedCycleNodes = cycleNodes.sort((a, b) => (a.id < b.id ? -1 : 1));

  const cycleNodeId = sortedCycleNodes.map((item) => item.id).join('-');

  cycleNodesMap.set(cycleNodeId, sortedCycleNodes);
}
