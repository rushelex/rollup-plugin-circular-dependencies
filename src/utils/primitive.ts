const toString = Object.prototype.toString;

const getTag = (value: unknown) => {
  if (value == null) {
    return value === undefined ? '[object Undefined]' : '[object Null]';
  }
  return toString.call(value);
};

export const isNil = (value: unknown): value is null | undefined => value == null;

export const isObject = (value: unknown): value is object => {
  const type = typeof value;
  return value != null && (type === 'object' || type === 'function');
};

export const isString = (value: unknown): value is string => {
  const type = typeof value;
  return (
    type === 'string' ||
    (type === 'object' && value != null && !Array.isArray(value) && getTag(value) === '[object String]')
  );
};
