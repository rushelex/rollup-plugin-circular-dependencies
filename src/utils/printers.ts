import { existsSync, mkdirSync, writeFileSync } from 'node:fs';
import { dirname } from 'node:path';

import type { Context } from '../context';

type FormattedData = any;

interface PrinterConstructorParams {
  context: Context;
}

abstract class Printer {
  protected ctx: Context;

  public constructor(params: PrinterConstructorParams) {
    this.ctx = params.context;
  }

  public shouldPrint(_data: FormattedData): boolean {
    return true;
  }

  public abstract print(data: FormattedData): void;
}

class FilePrinter extends Printer {
  public shouldPrint(): boolean {
    return Boolean(this.ctx.options.outputFilePath);
  }

  public print(data: FormattedData): void {
    const resultData = data ? String(data) : '';
    const dirPath = dirname(this.ctx.options.outputFilePath);

    if (!existsSync(dirPath)) {
      mkdirSync(dirPath, { recursive: true });
    }

    writeFileSync(this.ctx.options.outputFilePath, resultData, {
      encoding: 'utf-8',
    });
  }
}

class ConsolePrinter extends Printer {
  public shouldPrint(data: FormattedData): boolean {
    const isConsolePrinter = !this.ctx.options.outputFilePath;
    const isDataExists = !!data || data === 0 || (typeof data === 'string' && data.length > 0);

    return isConsolePrinter && isDataExists;
  }

  public print(data: FormattedData): void {
    // eslint-disable-next-line no-console
    console.info(`\n\n${String(data)}\n`);
  }
}

export function createPrinters(params: PrinterConstructorParams): Printer[] {
  const printers = [FilePrinter, ConsolePrinter];
  return printers.map((PrinterConstructor) => new PrinterConstructor(params));
}
