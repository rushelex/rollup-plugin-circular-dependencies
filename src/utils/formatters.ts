import chalk from 'chalk';

import { type CircularDependenciesData } from '../types';

import { isNil, isObject, isString } from './primitive';

type Formatter = (data: CircularDependenciesData) => string;

interface PrettyFormatterConfig {
  colors?: boolean;
}

function JSONFormatter(): Formatter {
  return (data): string => {
    if (!Array.isArray(data) && !isObject(data)) {
      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
      return data ?? '';
    }

    return JSON.stringify(data, null, 2);
  };
}

function PrettyFormatter(config?: PrettyFormatterConfig): Formatter {
  const { colors = true } = config || {};

  return (data): string => {
    if (!isNil(data) && isObject(data) && Object.keys(data).length === 0) {
      return '';
    }

    if (!isValidData(data)) {
      throw new Error('Pretty formatter can only be used with original output');
    }

    const groups = [];

    for (const item of Object.entries(data)) {
      const [entryModuleId, moduleNodes] = item;

      let group = '';

      group += colors ? chalk.yellow(entryModuleId) : entryModuleId;

      for (const currentCir of moduleNodes) {
        group += '\n' + '    ' + currentCir.join(colors ? chalk.blue(' -> ') : ' -> ');
      }

      groups.push(group);
    }

    return groups.join('\n\n');
  };
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isValidData(data: any): data is CircularDependenciesData {
  if (data && isObject(data)) {
    const [firstValue] = Object.values(data);

    if (Array.isArray(firstValue)) {
      const [firstChildValue] = firstValue;

      if (Array.isArray(firstChildValue) && firstChildValue.every(isString)) {
        return true;
      }
    }
  }

  return false;
}

export const DefaultFormatters = {
  JSON: JSONFormatter,
  Pretty: PrettyFormatter,
};
