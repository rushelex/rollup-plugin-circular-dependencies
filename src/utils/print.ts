import type { PluginContext } from 'rollup';

import type { Context } from '../context';
import type { ModuleNode } from '../module';
import type { CircularDependenciesData } from '../types';

import { createPrinters } from './printers';

const ERROR_MESSAGE = 'Circular dependencies has been detected';

export function printCycleNodes(ctx: Context, cycleNodes: Map<string, ModuleNode[]>, pluginContext: PluginContext) {
  const rawData = getRawData(cycleNodes);

  // eslint-disable-next-line ts/no-unsafe-assignment
  const formattedData = ctx.options.formatOut(transformNodePaths(ctx, rawData));

  ctx.options.onEnd(
    {
      rawOutput: rawData,
      // eslint-disable-next-line ts/no-unsafe-assignment
      formattedOutput: formattedData,
    },
    pluginContext,
  );

  for (const printer of createPrinters({ context: ctx })) {
    if (printer.shouldPrint(formattedData)) {
      printer.print(formattedData);
    }
  }

  validateCycleData(rawData, ctx, pluginContext);
}

function getRawData(data: Map<string, ModuleNode[]>) {
  return groupByFirstNodePath(getNodePaths(filterNodes(data)));
}

function filterNodes(cycleNodesMap: Map<string, ModuleNode[]>) {
  return Array.from(cycleNodesMap.values()).filter((item) => item.length);
}

function getNodePaths(data: ModuleNode[][]): Array<Array<ModuleNode['id']>> {
  return data.map((nodeArr) => {
    return nodeArr.map((item) => {
      return item.id;
    });
  });
}

function groupByFirstNodePath(data: Array<Array<ModuleNode['id']>>) {
  return data.reduce<CircularDependenciesData>((acc, curNodes) => {
    const firstNodeId = curNodes[0];

    acc[firstNodeId] ??= [];

    acc[firstNodeId].push(curNodes);

    return acc;
  }, {});
}

function transformNodePaths(ctx: Context, data: CircularDependenciesData): CircularDependenciesData {
  const transformedData: CircularDependenciesData = {};

  for (const [nodeId, cycles] of Object.entries(data)) {
    transformedData[ctx.options.formatOutModulePath(nodeId)] = cycles.map((cycle) => {
      return cycle.map((nodeId) => {
        return ctx.options.formatOutModulePath(nodeId);
      });
    });
  }

  return transformedData;
}

function validateCycleData(data: CircularDependenciesData, ctx: Context, pluginContext: PluginContext) {
  const handler = ctx.options.throwOnError ? pluginContext.error : pluginContext.warn;

  if (Object.keys(data).length > 0) {
    handler(ERROR_MESSAGE);
  }
}
