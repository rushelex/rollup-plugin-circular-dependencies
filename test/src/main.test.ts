import { existsSync, readFileSync, rmSync } from 'node:fs';
import { resolve } from 'node:path';
import { cwd } from 'node:process';
import { expect } from 'vitest';

import { getRollupRunningProcess, OUTPUT_FILE_RELATIVE_PATH } from './utils';

const CIRCULAR_DEPS_ERROR = 'Circular dependencies has been detected';

const testOutputFilePath = resolve(cwd(), OUTPUT_FILE_RELATIVE_PATH);

describe('main test', () => {
  beforeEach(() => {
    if (existsSync(testOutputFilePath)) {
      rmSync(testOutputFilePath);
    }
  });

  it('should print basic output in console', () => {
    const { stdout, stderr } = getRollupRunningProcess('output.console.basic');

    expect(stdout.toString()).toMatchSnapshot();

    expect(stderr.toString().includes(CIRCULAR_DEPS_ERROR)).toBeTruthy();
  });

  it('should print json output in console', () => {
    const { stdout, stderr } = getRollupRunningProcess('output.console.json');

    expect(stdout.toString()).toMatchSnapshot();

    expect(stderr.toString().includes(CIRCULAR_DEPS_ERROR)).toBeTruthy();
  });

  it('should print basic output in file', () => {
    const { stderr } = getRollupRunningProcess('output.file.basic');

    const outputContent = readFileSync(testOutputFilePath, { encoding: 'utf-8' });

    expect(outputContent).toMatchSnapshot();

    expect(stderr.toString().includes(CIRCULAR_DEPS_ERROR)).toBeTruthy();
  });

  it('should print pretty colored output in file', () => {
    const { stderr } = getRollupRunningProcess('output.file.pretty.colored');

    const outputContent = readFileSync(testOutputFilePath, { encoding: 'utf-8' });

    expect(outputContent).toMatchSnapshot();

    expect(stderr.toString().includes(CIRCULAR_DEPS_ERROR)).toBeTruthy();
  });

  it('should print pretty uncolored output in file', () => {
    const { stderr } = getRollupRunningProcess('output.file.pretty.uncolored');

    const outputContent = readFileSync(testOutputFilePath, { encoding: 'utf-8' });

    expect(outputContent).toMatchSnapshot();

    expect(stderr.toString().includes(CIRCULAR_DEPS_ERROR)).toBeTruthy();
  });

  it('should skip detecting', () => {
    const { stdout, stderr, status } = getRollupRunningProcess('disabled');

    expect(stdout.toString()).toMatchSnapshot();

    expect(stderr.toString().includes(CIRCULAR_DEPS_ERROR)).toBeFalsy();

    expect(status).toBe(0);
  });

  it('should exit with code 0', () => {
    const { status } = getRollupRunningProcess('status.warn');

    expect(status).toBe(0);
  });

  it('should exit with code 1', () => {
    const { status } = getRollupRunningProcess('status.error');

    expect(status).toBe(1);
  });

  it('should call lifecycle callbacks', () => {
    const { stdout } = getRollupRunningProcess('callbacks');

    expect(stdout.toString()).toMatchSnapshot();
  });
});
