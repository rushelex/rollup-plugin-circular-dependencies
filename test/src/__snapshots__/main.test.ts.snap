// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`main test > should call lifecycle callbacks 1`] = `
"[circularDependencies] onStart
[circularDependencies] onDetected | modulePath :>> src/fixtures/data/async_module/circleDep/data/data1.js
[circularDependencies] onDetected | modulePath :>> src/fixtures/data/async_module/depSelf/depSelf.js
[circularDependencies] onDetected | modulePath :>> src/fixtures/data/async_module/indirectCircleDep/data/data1.js
[circularDependencies] onDetected | modulePath :>> src/fixtures/data/sync_module/circleDep/circleDep1.js
[circularDependencies] onDetected | modulePath :>> src/fixtures/data/sync_module/depSelf/depSelf1.js
[circularDependencies] onDetected | modulePath :>> src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js
[circularDependencies] onEnd | rawOutput :>> {
  'src/fixtures/data/async_module/circleDep/data/data1.js': [
    [
      'src/fixtures/data/async_module/circleDep/data/data1.js',
      'src/fixtures/data/async_module/circleDep/data/data2.js'
    ],
    [
      'src/fixtures/data/async_module/circleDep/data/data1.js',
      'src/fixtures/data/async_module/circleDep/data/data3.js'
    ]
  ],
  'src/fixtures/data/async_module/depSelf/depSelf.js': [ [ 'src/fixtures/data/async_module/depSelf/depSelf.js' ] ],
  'src/fixtures/data/async_module/indirectCircleDep/data/data1.js': [
    [
      'src/fixtures/data/async_module/indirectCircleDep/data/data1.js',
      'src/fixtures/data/async_module/indirectCircleDep/data/data2.js',
      'src/fixtures/data/async_module/indirectCircleDep/data/data3.js'
    ]
  ],
  'src/fixtures/data/sync_module/circleDep/circleDep1.js': [
    [
      'src/fixtures/data/sync_module/circleDep/circleDep1.js',
      'src/fixtures/data/sync_module/circleDep/circleDep2.js'
    ],
    [
      'src/fixtures/data/sync_module/circleDep/circleDep1.js',
      'src/fixtures/data/sync_module/circleDep/circleDep3.js'
    ]
  ],
  'src/fixtures/data/sync_module/depSelf/depSelf1.js': [ [ 'src/fixtures/data/sync_module/depSelf/depSelf1.js' ] ],
  'src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js': [
    [
      'src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js',
      'src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep2.js',
      'src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep3.js'
    ]
  ]
}
[circularDependencies] onEnd | formattedOutput :>> [33msrc/fixtures/data/async_module/circleDep/data/data1.js[39m
    src/fixtures/data/async_module/circleDep/data/data1.js[34m -> [39msrc/fixtures/data/async_module/circleDep/data/data2.js
    src/fixtures/data/async_module/circleDep/data/data1.js[34m -> [39msrc/fixtures/data/async_module/circleDep/data/data3.js

[33msrc/fixtures/data/async_module/depSelf/depSelf.js[39m
    src/fixtures/data/async_module/depSelf/depSelf.js

[33msrc/fixtures/data/async_module/indirectCircleDep/data/data1.js[39m
    src/fixtures/data/async_module/indirectCircleDep/data/data1.js[34m -> [39msrc/fixtures/data/async_module/indirectCircleDep/data/data2.js[34m -> [39msrc/fixtures/data/async_module/indirectCircleDep/data/data3.js

[33msrc/fixtures/data/sync_module/circleDep/circleDep1.js[39m
    src/fixtures/data/sync_module/circleDep/circleDep1.js[34m -> [39msrc/fixtures/data/sync_module/circleDep/circleDep2.js
    src/fixtures/data/sync_module/circleDep/circleDep1.js[34m -> [39msrc/fixtures/data/sync_module/circleDep/circleDep3.js

[33msrc/fixtures/data/sync_module/depSelf/depSelf1.js[39m
    src/fixtures/data/sync_module/depSelf/depSelf1.js

[33msrc/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js[39m
    src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js[34m -> [39msrc/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep2.js[34m -> [39msrc/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep3.js


[33msrc/fixtures/data/async_module/circleDep/data/data1.js[39m
    src/fixtures/data/async_module/circleDep/data/data1.js[34m -> [39msrc/fixtures/data/async_module/circleDep/data/data2.js
    src/fixtures/data/async_module/circleDep/data/data1.js[34m -> [39msrc/fixtures/data/async_module/circleDep/data/data3.js

[33msrc/fixtures/data/async_module/depSelf/depSelf.js[39m
    src/fixtures/data/async_module/depSelf/depSelf.js

[33msrc/fixtures/data/async_module/indirectCircleDep/data/data1.js[39m
    src/fixtures/data/async_module/indirectCircleDep/data/data1.js[34m -> [39msrc/fixtures/data/async_module/indirectCircleDep/data/data2.js[34m -> [39msrc/fixtures/data/async_module/indirectCircleDep/data/data3.js

[33msrc/fixtures/data/sync_module/circleDep/circleDep1.js[39m
    src/fixtures/data/sync_module/circleDep/circleDep1.js[34m -> [39msrc/fixtures/data/sync_module/circleDep/circleDep2.js
    src/fixtures/data/sync_module/circleDep/circleDep1.js[34m -> [39msrc/fixtures/data/sync_module/circleDep/circleDep3.js

[33msrc/fixtures/data/sync_module/depSelf/depSelf1.js[39m
    src/fixtures/data/sync_module/depSelf/depSelf1.js

[33msrc/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js[39m
    src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js[34m -> [39msrc/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep2.js[34m -> [39msrc/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep3.js

"
`;

exports[`main test > should print basic output in console 1`] = `
"

[33msrc/fixtures/data/async_module/circleDep/data/data1.js[39m
    src/fixtures/data/async_module/circleDep/data/data1.js[34m -> [39msrc/fixtures/data/async_module/circleDep/data/data2.js
    src/fixtures/data/async_module/circleDep/data/data1.js[34m -> [39msrc/fixtures/data/async_module/circleDep/data/data3.js

[33msrc/fixtures/data/async_module/depSelf/depSelf.js[39m
    src/fixtures/data/async_module/depSelf/depSelf.js

[33msrc/fixtures/data/async_module/indirectCircleDep/data/data1.js[39m
    src/fixtures/data/async_module/indirectCircleDep/data/data1.js[34m -> [39msrc/fixtures/data/async_module/indirectCircleDep/data/data2.js[34m -> [39msrc/fixtures/data/async_module/indirectCircleDep/data/data3.js

[33msrc/fixtures/data/sync_module/circleDep/circleDep1.js[39m
    src/fixtures/data/sync_module/circleDep/circleDep1.js[34m -> [39msrc/fixtures/data/sync_module/circleDep/circleDep2.js
    src/fixtures/data/sync_module/circleDep/circleDep1.js[34m -> [39msrc/fixtures/data/sync_module/circleDep/circleDep3.js

[33msrc/fixtures/data/sync_module/depSelf/depSelf1.js[39m
    src/fixtures/data/sync_module/depSelf/depSelf1.js

[33msrc/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js[39m
    src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js[34m -> [39msrc/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep2.js[34m -> [39msrc/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep3.js

"
`;

exports[`main test > should print basic output in file 1`] = `
"{
  "src/fixtures/data/async_module/circleDep/data/data1.js": [
    [
      "src/fixtures/data/async_module/circleDep/data/data1.js",
      "src/fixtures/data/async_module/circleDep/data/data2.js"
    ],
    [
      "src/fixtures/data/async_module/circleDep/data/data1.js",
      "src/fixtures/data/async_module/circleDep/data/data3.js"
    ]
  ],
  "src/fixtures/data/async_module/depSelf/depSelf.js": [
    [
      "src/fixtures/data/async_module/depSelf/depSelf.js"
    ]
  ],
  "src/fixtures/data/async_module/indirectCircleDep/data/data1.js": [
    [
      "src/fixtures/data/async_module/indirectCircleDep/data/data1.js",
      "src/fixtures/data/async_module/indirectCircleDep/data/data2.js",
      "src/fixtures/data/async_module/indirectCircleDep/data/data3.js"
    ]
  ],
  "src/fixtures/data/sync_module/circleDep/circleDep1.js": [
    [
      "src/fixtures/data/sync_module/circleDep/circleDep1.js",
      "src/fixtures/data/sync_module/circleDep/circleDep2.js"
    ],
    [
      "src/fixtures/data/sync_module/circleDep/circleDep1.js",
      "src/fixtures/data/sync_module/circleDep/circleDep3.js"
    ]
  ],
  "src/fixtures/data/sync_module/depSelf/depSelf1.js": [
    [
      "src/fixtures/data/sync_module/depSelf/depSelf1.js"
    ]
  ],
  "src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js": [
    [
      "src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js",
      "src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep2.js",
      "src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep3.js"
    ]
  ]
}"
`;

exports[`main test > should print json output in console 1`] = `
"

{
  "src/fixtures/data/async_module/circleDep/data/data1.js": [
    [
      "src/fixtures/data/async_module/circleDep/data/data1.js",
      "src/fixtures/data/async_module/circleDep/data/data2.js"
    ],
    [
      "src/fixtures/data/async_module/circleDep/data/data1.js",
      "src/fixtures/data/async_module/circleDep/data/data3.js"
    ]
  ],
  "src/fixtures/data/async_module/depSelf/depSelf.js": [
    [
      "src/fixtures/data/async_module/depSelf/depSelf.js"
    ]
  ],
  "src/fixtures/data/async_module/indirectCircleDep/data/data1.js": [
    [
      "src/fixtures/data/async_module/indirectCircleDep/data/data1.js",
      "src/fixtures/data/async_module/indirectCircleDep/data/data2.js",
      "src/fixtures/data/async_module/indirectCircleDep/data/data3.js"
    ]
  ],
  "src/fixtures/data/sync_module/circleDep/circleDep1.js": [
    [
      "src/fixtures/data/sync_module/circleDep/circleDep1.js",
      "src/fixtures/data/sync_module/circleDep/circleDep2.js"
    ],
    [
      "src/fixtures/data/sync_module/circleDep/circleDep1.js",
      "src/fixtures/data/sync_module/circleDep/circleDep3.js"
    ]
  ],
  "src/fixtures/data/sync_module/depSelf/depSelf1.js": [
    [
      "src/fixtures/data/sync_module/depSelf/depSelf1.js"
    ]
  ],
  "src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js": [
    [
      "src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js",
      "src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep2.js",
      "src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep3.js"
    ]
  ]
}

"
`;

exports[`main test > should print pretty output in file 1`] = `
"src/fixtures/data/async_module/circleDep/data/data1.js
    src/fixtures/data/async_module/circleDep/data/data1.js -> src/fixtures/data/async_module/circleDep/data/data2.js
    src/fixtures/data/async_module/circleDep/data/data1.js -> src/fixtures/data/async_module/circleDep/data/data3.js

src/fixtures/data/async_module/depSelf/depSelf.js
    src/fixtures/data/async_module/depSelf/depSelf.js

src/fixtures/data/async_module/indirectCircleDep/data/data1.js
    src/fixtures/data/async_module/indirectCircleDep/data/data1.js -> src/fixtures/data/async_module/indirectCircleDep/data/data2.js -> src/fixtures/data/async_module/indirectCircleDep/data/data3.js

src/fixtures/data/sync_module/circleDep/circleDep1.js
    src/fixtures/data/sync_module/circleDep/circleDep1.js -> src/fixtures/data/sync_module/circleDep/circleDep2.js
    src/fixtures/data/sync_module/circleDep/circleDep1.js -> src/fixtures/data/sync_module/circleDep/circleDep3.js

src/fixtures/data/sync_module/depSelf/depSelf1.js
    src/fixtures/data/sync_module/depSelf/depSelf1.js

src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js
    src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep1.js -> src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep2.js -> src/fixtures/data/sync_module/indirectCircleDep/indirectCircleDep3.js"
`;

exports[`main test > should skip detecting 1`] = `""`;
